src/clr-angular/data/datagrid/providers/columns.service.ts

```diff
@@ -1,15 +1,38 @@
 /*
- * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
  * This software is released under MIT license.
  * The full license information can be found in LICENSE in the root directory of this project.
  */
 import { Injectable } from '@angular/core';
 import { BehaviorSubject } from 'rxjs';
 import { ColumnStateDiff, ColumnState } from '../interfaces/column-state.interface';
+import { ALL_COLUMN_CHANGES } from '../enums/column-changes.enum';
 
 @Injectable()
 export class ColumnsService {
   columns: BehaviorSubject<ColumnState>[] = [];
+  private _cache: ColumnState[];
+
+  cache() {
+    this._cache = this.columns.map(subject => {
+      const value = subject.value;
```

Comment:
I think we should have a new object like this:
```
const value = { ...subject.value };
```
Otherwise, the cached state objects still have reference to the actual column states. If they change, the cached objects will change.

Reply:
Yes, should fix up.